src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,217 @@
+        if (size == 0) {
+            addFirstElement(newNode);
+            return;
+        }
+        if (index == size || index == 0) {
+            addTopBottom(newNode, index);
+            return;
+        }
+        newNode.prev = oldNode.prev;
+        newNode.next = oldNode;
+        oldNode.prev.next = newNode;
+        oldNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexGetSet(index);
+        Node<T> node = getIndexNode(index);
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexGetSet(index);
+        Node<T> node = getIndexNode(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexGetSet(index);
+        Node<T> node = getIndexNode(index);
+        T removedItem = node.item;
+        if (index == 0) {
+            removeBottom();
+            return removedItem;
+        }
+        if (index == size - 1) {
+            removeTop();
+            return removedItem;
+        }
+        removeNode(node);
+        return removedItem;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = getNode(object);
+        if (node == null) {
+            return false;
+        }
+        if (node.prev == null) {
+            removeByObject();
+            return true;
+        }
+        if (node.next == null) {
+            removeTop();
+            return true;
+        }
+        removeNode(node);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private void checkIndexGetSet(int index) {
+        if (index + 1 > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private void addFirstElement(Node<T> newNode) {
+        if (size == 0) {
+            head = newNode;
+            tail = newNode;
+            size++;
+        }
+    }
+
+    private Node<T> getIndexNode(int index) {
+        Node<T> node;
+        if (index < (size >> 1)) {
+            node = head;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        }
+        node = tail;
+        for (int i = size - 1; i > index; i--) {
+            node = node.prev;
+        }
+        return node;
+    }
```

Comment:
It seemed that you reassign value of node variable after first loop in second loop. Maybe add else clause or one more return in if clause? Because if this statement (if (index < (size >> 1))) is true, it does job, but after that second loop do it again and change value of node.